### ISSUE TITLE
Support fetching all completed tasks

### ISSUE THREAD

**Issue by programmerPhysicist** [Doist/todoist-api-python#123]
## Question description
I get a HTTPError when I attempt to run `get_completed_items()` off the api object. I get the following error:
![ApiErrorTodoist](https://github.com/Doist/todoist-api-python/assets/20853454/d042e97a-edf5-4501-9c8f-1b1bfd258438)
is this a bug, or am I doing something wrong?

**Comment by brunorosilva** [comment]
Hi, you seem to be using `get_completed_items()` without any project nor section filter, which is not supported by this api version (insane, I know). They should give other options to get all completed items, as it was possible before the version update.

### Call to endpoint without project_id field
![image](https://github.com/Doist/todoist-api-python/assets/17326742/6ec47d6c-1e2b-47f9-9776-17141d727376)

### Call to endpoint with project_id field
![image](https://github.com/Doist/todoist-api-python/assets/17326742/64edbe28-5274-4200-a523-35cd1138cce0)

There's no direct fix to this, but there's a workaround for non-archived projects. If you know all of the project_ids you can call it with `get_all_completed_items(project_id="1231924012")`.

@lefcha do you guys have any plans to add back the call without the `project_id` filter? Or maybe give us a way to get archived projects.

**Comment by lefcha** [comment]
Yes, while the public API has a way to get all completed tasks, ie.

https://developer.todoist.com/sync/v9/#get-all-completed-items

unfortunately we haven't yet implemented very much from the auxiliary functions of the Sync API in this SDK. Your best option for now would be to directly send a request to the `https://api.todoist.com/sync/v9/completed/get_all` endpoint.

There's also an API endpoint to get archived projects `https://api.todoist.com/sync/v9/projects/get_archived`, in case you find it useful, documented at:

https://developer.todoist.com/sync/v9/#get-archived-projects

This SDK is primarily about implementing the REST API, and we don't plan to implement all the auxiliary functions of the Sync API at the moment.

**Comment by oscarclivio** [comment]
Same problem here.

**Comment by brunorosilva** [comment]
@oscarclivio did you try to call the `get_completed_items()` using a `project_id` argument?

If you really want to get all of the completed items or you do not know the project_ids you're supposed to use, use the sync api - as lefcha answered me. If you're not sure how to do this, feel free to use this code on my working branch [code](https://github.com/brunorosilva/todoist-analytics/blob/fix/api-version-bump/todoist_analytics/backend/sync_api.py)

**Comment by oscarclivio** [comment]
Thanks @brunorosilva for helping. I do yes and I iterate over project ids. As every call of `get_completed_items()` using a `project_id` returns a limited number of result, I iterate calls, and given the large number of completed tasks I stop doing calls for a fixed `project_id` when I get tasks that are before a certain date threshold. However it seems that only root (non-child) complete tasks are returned.... Do you have any insight on this?

**Comment by oscarclivio** [comment]
Update :  @brunorosilva I have implemented your solution. It does return all **root** completed items from all projects, but not any completed item that has a parent! Do you have any clues on how to access these completed items?

**Comment by brunorosilva** [comment]
@oscarclivio Glad I could help!

Yeah, that happens because [their implementation of `get_completed_items()`](https://github.com/Doist/todoist-api-python/blob/main/todoist_api_python/api.py#L213) calls to `https://api.todoist.com/sync/v9/archive/items?project_id=<project_id>`, and each call takes ~500ms to get 100 completed items (the limit for each call) for a single project as the example below. The endpoint I'm using is `https://api.todoist.com/sync/v9/completed/get_all`, which takes ~700ms to get 200 completed items (the limit), so this is why it's faster. 

Their approach is not wrong, it is in line with their [argument](https://developer.todoist.com/guides/#our-apis) for having two APIs. Rest providing simpler use cases and Sync for data heavy use cases.

Below are some images and comparisons I made.
![image](https://github.com/Doist/todoist-api-python/assets/17326742/ec70a8b1-9cf8-4926-9c68-a7bcb1999100)

Using /archive
See the time below for one project -> 570ms
![Screenshot from 2023-11-24 16-23-37](https://github.com/Doist/todoist-api-python/assets/17326742/2aeaf2f6-66b6-4435-8c8b-32a169b98987)
Using /completed
![image](https://github.com/Doist/todoist-api-python/assets/17326742/b746a5b8-699d-4e09-9aeb-533adb037218)

**Comment by programmerPhysicist** [comment]
I came up with the following implementation:
```
def dict_to_Task(obj, url):
    obj['comment_count'] = obj['note_count']
    obj['is_completed'] = (obj['completed_at'] != '')
    obj['created_at'] = "unknown"
    obj['creator_id'] = obj['user_id']
    obj['description'] = obj['content']
    obj['priority'] = ''
    obj['url'] = url
    return Task.from_dict(obj)

def get_all_completed_items(api):
    url = get_sync_url('completed/get_all')
    completed_items = get(_session, url, _token)
    tasks = completed_items['items']
    return [dict_to_Task(obj, url) for obj in tasks]
```
This probably could be added to the api.

**Comment by goncalossilva** [comment]
Hey, all! Our SDK is currently working as intended, but I'll repurpose this issue as a feature request for ways to get _all_ completed tasks. It would indeed be a welcome addition to it.
